
Theory of measurement for site-specific evolutionary rates in amino-acid sequences
==================================================================================

This repository contains all the scripts and data to reproduce the results of:

D. K. Sydykova, C. O. Wilke (2018). Theory of measurement for site-specific evolutionary rates in amino-acid sequences

Contents
--------

-   `analytical_rates` contains rates that were calculated using analytical derivations for site-wise evolutionary rate. The files in this direcotry are:
    -   `all_sites_aa.csv`
    -   `ten_sites_aa.csv`
    -   `ten_sites_aa_true_JC.csv`
    -   `codon_aa_equil_freq.txt`
    -   `ten_sites_aa_QM.csv`
    -   `ten_sites_codon.csv`
-   `inferred_rates`
-   `q_matrix` contains site-wise substitution matrices *Q* used for simulating alignments.
    -   `amino_acid` contains amino acid substitution matrices
    -   `codon` contains codon substitution matrice

    Files that start with `132L_A` indicate substitution matrices that were calculated using data from Echave et al (2015) for egg white lysozyme (PDB ID: 132L). For example, file `132L_A_site79_q_matrix.npy` corresponds to the substitution matrix calculated for site 79. The site positions here correspond to the site positions given by the file `132L_A_foldx_ddG.txt`, which was directly copied from the git repository for Echave et al (2015) [https://github.com/wilkelab/therm\_constraints\_rate\_variation](https://github.com/wilkelab/therm_constraints_rate_variation/tree/master/ddG_calculations/foldX/foldx_ddG). These matrices were calculated according to the mutation-selection (MutSel) theory by Halpern and Bruno (1998). The script to calculate amino acid MutSel matrices is `src/calculate_aa_mutsel_Q.py`, and the script to calculate codon MutSel matrices is `src/calculate_codon_mutsel_Q.py`.

    Files that start with `site0_JC` indicate substitution matrices *Q* defined as *Q* = *r*<sup>(*k*)</sup>*Q*<sub>JC</sub>. Here, *Q*<sub>JC</sub> is the Jukes-Cantor-like matrix, and *r*<sup>(*k*)</sup> is the true rate at site *k*. True rates were generated by the script `src/analytical_rate_aa_true_JC.py` and were stored in `analytical_rates/ten_sites_aa_true_JC.csv`. The script to calculate amino acid substitution matrices *Q* defined as *Q* = *r*<sup>(*k*)</sup>*Q*<sub>JC</sub> is `src/calculate_aa_true_JC_Q.py`.

-   `trees`
-   `hyphy`
-   `plots`
-   `src`

Analysis
--------
